snapcraft/internal/build_providers/_lxd/_lxd.py

```diff
@@ -0,0 +1,277 @@
+        raise NotImplementedError(
+            "Feature not supported with provider {!r}".format(self._get_provider_name())
+        )
+
+    def _push_file(self, *, source: str, destination: str) -> None:
+        self._ensure_container_running()
+
+        # TODO: better handling of larger files.
+        with open(source, "wb") as source_data:
+            try:
+                self._container.files.put(destination, source_data.read())
+            except pylxd.exceptions.LXDAPIException as lxd_api_error:
+                raise errors.ProviderFileCopyError(
+                    provider_name=self._get_provider_name(), error_message=lxd_api_error
+                )
+
+    def __init__(self, *, project, echoer, is_ephemeral: bool = False) -> None:
+        super().__init__(project=project, echoer=echoer, is_ephemeral=is_ephemeral)
+        self.echoer.warning(
+            "The LXD provider is offered as a technology preview for early adopters. "
+            "The command line interface, container names or lifecycle handling may "
+            "change in upcoming releases."
+        )
+        self._lxd_client = pylxd.Client()
+        self._container = None  # type: Optional[pylxd.models.container.Container]
+
+    def create(self) -> None:
+        """Create the LXD instance and setup the build environment."""
+        self.launch_instance()
+
+    def destroy(self) -> None:
+        """Destroy the instance, trying to stop it first."""
+        self._stop()
+
+    def mount_project(self) -> None:
+        if self._PROJECT_DEVICE_NAME in self._container.devices:
+            return
+
+        # Resolve the home directory
+        home_dir = (
+            self._run(command=["printenv", "HOME"], hide_output=True).decode().strip()
+        )
+
+        self._container.sync()
+
+        self._container.devices[self._PROJECT_DEVICE_NAME] = {
+            "type": "disk",
+            "source": self.project._project_dir,
+            "path": os.path.join(home_dir, "project"),
+        }
+
+        try:
+            self._container.save(wait=True)
+        except pylxd.exceptions.LXDAPIException as lxd_api_error:
+            raise errors.ProviderMountError(
+                provider_name=self._get_provider_name(), error_message=lxd_api_error
+            ) from lxd_api_error
+
+    def clean_project(self) -> bool:
+        was_cleaned = super().clean_project()
+        if was_cleaned:
+            if self._container is None:
+                self._container = self._lxd_client.containers.get(self.instance_name)
+            self._stop()
+            self._container.delete(wait=True)
+        return was_cleaned
+
+    def pull_file(self, name: str, destination: str, delete: bool = False) -> None:
+        self._ensure_container_running()
+
+        # TODO: better handling of larger files.
+        try:
+            source_data = self._container.files.get(name)
+        except pylxd.exceptions.LXDAPIException as lxd_api_error:
```

Comment:
If pylxd.exceptions.LXDAPIException happens it's a bug in LXD, right? Should we instruct them to file one?

Reply:
This can be anything, not necessarily a bug... They only have one exception so we cannot know exactly what is going on, but they use the stdlib exceptions for anything not return from the lxd socket.

Examples of this are:
- Failed to reach the lxd remote
- Failed to create the container
- Failed to save the config

Think of it as multipass exiting with != 0